File: solutions/day3.py
def get_data(day):
    text_file = open('../data/day' + str(day) + ".txt", "r")
    input_data = text_file.read()
    text_file.close()
    return input_data


def data_prep(day):
    input_data = get_data(day)
    input_data = input_data.split('\n')
    input_data = [i.split()[0] for i in input_data]
    input_data = [list(map(int, i)) for i in input_data]

    return input_data


def get_dec(number_list):
    return int("".join(str(i) for i in number_list), 2)


def part1(data):
    counts = []
    for j in range(len(data[0])):
        counts.append(sum(i[j] for i in data))

    counts = [i / len(data) for i in counts]
    gamma_bin = [1 * (i >= 0.5) for i in counts]
    epsilon_bin = [1 * (i < 0.5) for i in counts]

    power_consumption = get_dec(gamma_bin) * get_dec(epsilon_bin)

    return power_consumption


def part2(data):
    j = 0
    oxygen_data = data.copy()
    while j < len(oxygen_data[0]) and len(oxygen_data) > 1:
        criteria = 1 * ((sum(i[j] for i in oxygen_data) / len(oxygen_data)) >= 0.5)
        new_data = []
        for i in oxygen_data:
            if i[j] == criteria:
                new_data.append(i)

        oxygen_data = new_data
        j += 1

    oxygen_dec = get_dec(oxygen_data[0])

    j = 0
    carbon_data = data.copy()
    while j < len(carbon_data[0]) and len(carbon_data) > 1:
        criteria = 1 * ((sum(i[j] for i in carbon_data) / len(carbon_data)) < 0.5)
        new_data = []
        for i in carbon_data:
            if i[j] == criteria:
                new_data.append(i)

        carbon_data = new_data
        j += 1

    carbon_dec = get_dec(carbon_data[0])

    return oxygen_dec * carbon_dec


def main():
    day = 3
    input_data = data_prep(day)

    answer1 = part1(input_data)
    answer2 = part2(input_data)
    print(f'Day {day} part 1 answer: {answer1}')
    print(f'Day {day} part 2 answer: {answer2}')


if __name__ == "__main__":
    main()

File: go_solutions/day3.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {

	fileName := "data/day3.txt"
	input := getData(fileName)

	var data [][]string

	for _, line := range input {
		s := strings.Split(line, "")
		data = append(data, s)
	}

	fmt.Println("Part 1 answer:", part1(data))
	fmt.Println("Part 2 answer:", part2(data))

}

func getData(f string) []string {
	file, err := os.Open(f)
	if err != nil {
		fmt.Println(err)
	}

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines
}

func mostCommonDigit(data [][]string, index int) int {
	count := 0
	for _, n := range data {
		if n[index] == "1" {
			count++
		}
	}
	if count > len(data)/2 {
		return 1
	} else {
		return 0
	}
}

func binaryStringToDecimalInt(bin string) int {
	dec := 0
	exponent := 0
	for i, n := range bin {
		exponent = len(bin) - 1 - i
		if n == 49 {
			dec += int(math.Pow(2, float64(exponent)))
		}
	}
	return dec
}

func stringListToBinString(list []string) string {
	bin := ""
	for _, n := range list {
		if n == "1" {
			bin += n
		} else {
			bin += "0"
		}
	}
	return bin
}

func findTarget(data [][]string, moreOrLess string) [][]string {
	index := 0
	criteria := ""
	for index < 12 && len(data) > 1 {
		var newData [][]string
		if moreOrLess == "more" {
			criteria = strconv.Itoa(mostCommonDigit(data, index))
		} else {
			criteria = strconv.Itoa(1 - mostCommonDigit(data, index))
		}

		for _, n := range data {
			if n[index] == criteria {
				newData = append(newData, n)
			}
		}
		data = newData
		index++
	}
	return data
}

func part1(data [][]string) int {
	count := make([]int, 12)
	var gammaBin string
	var epsilonBin string

	for i, _ := range count {
		count[i] = mostCommonDigit(data, i)
		gammaBin += strconv.Itoa(count[i])
		epsilonBin += strconv.Itoa(1 - count[i])
	}

	powerConsumption := binaryStringToDecimalInt(gammaBin) * binaryStringToDecimalInt(epsilonBin)

	return powerConsumption
}

func part2(data [][]string) int {
	oxygenData := data
	carbonData := data

	oxygenData = findTarget(oxygenData, "more")
	oxygenBin := stringListToBinString(oxygenData[0])

	carbonData = findTarget(carbonData, "less")
	carbonBin := stringListToBinString(carbonData[0])

	return binaryStringToDecimalInt(oxygenBin) * binaryStringToDecimalInt(carbonBin)
}
